### PR TITLE
Adds `Function` custom resource

### PR DESCRIPTION
**Pull request by a-hilaly**
Issue: https://github.com/aws-controllers-k8s/community/issues/1915

Description of changes:
- Generate function resource
- Add necessary hooks
- Add e2e tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: apis/v1alpha1/generator.yaml [hunk 1]
```diff
@@ -56,3 +59,6 @@ resources:
         from:
           operation: CreateCachePolicy
           path: ETag
+  Function:
+    tags:
+      ignore: true
```
Comment (acornett21) [inline apis/v1alpha1/generator.yaml:64]: NIT: no new line

File: test/e2e/function.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    """Returns a dict containing the Function record from the CloudFront
+    API.
+    If no such Function exists, returns None.
+    """
+    c = boto3.client('cloudfront')
+    try:
+        resp = c.describe_function(Name=function_name)
+        return resp['Function']
+    except c.exceptions.NoSuchFunction:
+        return None
```
Comment (acornett21) [inline test/e2e/function.py:96]: NIT: no new line

File: test/e2e/resources/function.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        statusCode: 302,
+        statusDescription: 'Found',
+        headers: {
+          location: { value: 'https://${DOMAIN_NAME}/' }
+        }
+      }
+    }
+  functionConfig:
+    comment: function to redirect to $DOMAIN_NAME
+    runtime: $FUNCTION_RUNTIME
```
Comment (acornett21) [inline test/e2e/resources/function.yaml:19]: NIT: no new line

File: test/e2e/tests/test_function.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                    "comment": "New comment"
+                }
+            },
+        }
+        k8s.patch_custom_resource(ref, updates)
+        time.sleep(MODIFY_WAIT_AFTER_SECONDS)
+
+        latest = function.get(function_name)
+        assert latest is not None
+        assert latest['FunctionConfig']['Comment'] == "New comment"
```
Comment (acornett21) [inline test/e2e/tests/test_function.py:110]: NIT: no new line

File: pkg/resource/function/hooks.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		&svcsdk.GetFunctionInput{
+			Name:  r.Spec.Name,
+			Stage: r.Status.FunctionSummary.FunctionMetadata.Stage,
+		},
+	)
+	rm.metrics.RecordAPICall("GET", "GetFunction", err)
+	if err != nil {
+		return err
+	}
+	r.Spec.FunctionCode = []byte(output.FunctionCode)
```
Comment (a-hilaly) [inline pkg/resource/function/hooks.go:57]: I'm glad that the API returns something we can compare to the desired state, making very easy to update a function's code..

**Pull request merged**